compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -693,18 +696,51 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         let may_apply = match (source.kind(), target.kind()) {
             // Trait+Kx+'a -> Trait+Ky+'b (upcasts).
             (&ty::Dynamic(ref data_a, ..), &ty::Dynamic(ref data_b, ..)) => {
-                // Upcasts permit two things:
+                // Upcasts permit several things:
                 //
                 // 1. Dropping auto traits, e.g., `Foo + Send` to `Foo`
                 // 2. Tightening the region bound, e.g., `Foo + 'a` to `Foo + 'b` if `'a: 'b`
+                // 3. Tightening trait to its super traits, eg. `Foo` to `Bar` if `Foo: Bar`
                 //
-                // Note that neither of these changes requires any
-                // change at runtime. Eventually this will be
-                // generalized.
+                // Note that neither of the first two of these changes requires any
+                // change at runtime. The third needs to switch pointer metadata at runtime.
                 //
                 // We always upcast when we can because of reason
                 // #2 (region bounds).
-                data_a.principal_def_id() == data_b.principal_def_id()
+                let principal_compatible;
+                if data_a.principal_def_id() == data_b.principal_def_id() {
+                    principal_compatible = true;
+                } else {
+                    let is_principal_upcasting = match (data_a.principal(), data_b.principal()) {
+                        (Some(principal_a), Some(principal_b)) => {
+                            let trait_ref_a = principal_a
+                                .with_self_ty(self.tcx(), obligation.self_ty().skip_binder());
+                            let trait_ref_b = principal_b
+                                .with_self_ty(self.tcx(), obligation.self_ty().skip_binder());
+                            util::supertraits(self.tcx(), trait_ref_a)
+                                .any(|trait_ref| trait_ref == trait_ref_b)
```

Comment:
I'm not crazy about this `==`. That would rule out e.g. upcasting `Foo<?X>` to `Bar<Baz>`, even if unifying `?X` to `Baz` would work. I am debating what I think the *right* answer is here, though. Have to think through the case with ambigiuties. Probably we want something like "can match" and we want to wait to select a particular path until we have an unambiguous match.

This code is invoked after we've already decided that we *will* coerce, right? Or is it also used to test if we *should* coerce?

Reply:
Currently it is used to "early" test if we should test whether it should/could coerce (with obligations). 

I think maybe i should replicate the inner logic, using obligation engine to feature gate... Let me try it out.

Reply:
I'll also add a unit test for the example, thanks for the pointer!

Reply:
The implementation changed significantly, i'm resolving this conversion first.